### PR TITLE
Fixed #33253 -- Tightened regex matching JS module import/exports.

### PR DESCRIPTION
**Pull request by carltongibson**
WIP: Comments welcome. 

I added the highlight.js source which shows the wrong match from the `.*?` 
(It matches down to a `s={keyword:"do while select delete by update from",constant:"`, some pages of text.) 

I can reduce that down, putting in a `;` or similar... — not sure what makes a good test? 

Then `[_$a-zA-Z\*\s{},]` is roughly Most valid JS identifiers (plus `{}` and `\s`) as a best effort. 

Catching all identifiers isn't on I think: 

* https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Grammar_and_types
* https://mathiasbynens.be/notes/javascript-identifiers-es6

But open to ideas. 

Works with the existing cases. Any more we want to add. 
Thanks!

### CODE REVIEW

**Comment by kezabelle** [comment]
Are we looking to match as many import syntaxes as possible? Using the [example from MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import#syntax) and a rudimentary ipython playaround, it looks like we're missing some of them (despite those characters being there. I haven't analysed the regex properly for _why_, yet):

```
In [1]: tests = """import defaultExport from "module-name";
   ...: import * as name from "module-name";
   ...: import { export1 } from "module-name";
   ...: import { export1 as alias1 } from "module-name";
   ...: import { export1 , export2 } from "module-name";
   ...: import { export1 , export2 as alias2 } from "module-name";
   ...: import defaultExport, { export1 } from "module-name";
   ...: import defaultExport, * as name from "module-name";
   ...: """.splitlines()

In [2]: import re

In [3]: regex = re.compile(r"""(?P<matched>import\s+(?s:(?P<imports>[_$a-zA-Z\*\s{},]*?))\s*from\s*["'](?P<url>.*?)["'])""", re.IGNORECASE)

In [4]: for test in tests:
   ...:     result = regex.match(test)
   ...:     if result:
   ...:         print(result.groupdict())
   ...:     else:
   ...:         print('failed', test)
   ...:
{'matched': 'import defaultExport from "module-name"', 'imports': 'defaultExport', 'url': 'module-name'}
{'matched': 'import * as name from "module-name"', 'imports': '* as name', 'url': 'module-name'}
failed import { export1 } from "module-name";
failed import { export1 as alias1 } from "module-name";
failed import { export1 , export2 } from "module-name";
failed import { export1 , export2 as alias2 } from "module-name";
failed import defaultExport, { export1 } from "module-name";
{'matched': 'import defaultExport, * as name from "module-name"', 'imports': 'defaultExport, * as name', 'url': 'module-name'}
```
I've not tried the equivalent multi-line versions, but I'd guess it'd be the same.

**Comment by carltongibson** [comment]
I didn't have a `-`. 🤔

I'd say aiming for Best Reasonable Effort over As Many as Possible.

**Comment by kezabelle** [comment]
Ah that makes sense, I was misled by all of them containing curly braces even though you'd included those characters in the regex; the examples given all look to pass with the addition of `0-9`.

I was _initially_ going to suggest adding ID_START + ID_CONTINUE unicode ranges, because the JS ecosystem is full of small packages, and it'd be nice to support those imports/exports (even if they're a '_joke_' name like `ಠ_ಠ`) but then I looked at JS implementation for [ID_START](https://github.com/node-unicode/unicode-9.0.0/blob/main/Binary_Property/ID_Start/regex.js) and [ID_CONTINUE](https://github.com/node-unicode/unicode-9.0.0/blob/main/Binary_Property/ID_Continue/regex.js) and it seems like an exceptionally high bar to reach for :) I wonder if the Python stdlib contains the character classes somewhere; IIRC the rules are much the same, so if they could be _generated_ it might be more feasible to consider. Otherwise, restricting to an unfortunately 'common' ASCII extraction is probably the best effort, though it may mean a drip-feed of 'exotic' characters tickets over time...

**Comment by carltongibson** [comment]
This is the sort of topic where I might ask @shaib if he had any wisdom… (😉)

File: tests/staticfiles_tests/project/documents/cached/module.js [hunk 1]
```diff
@@ -5,8 +5,8 @@ import * as NewModule from "./module_test.js";
 import { testConst as alias } from "./module_test.js";
 import { firstConst, secondConst } from "./module_test.js";
 import {
-    firstVar as firstVarAlias,
-    secondVar as secondVarAlias
+    firstVar1 as firstVarAlias,
+    secondVar2 as secondVarAlias
```
Comment (felixxm) [inline tests/staticfiles_tests/project/documents/cached/module.js:9]:
```suggestion
    $second_var_2 as secondVarAlias
```

File: django/contrib/staticfiles/storage.py [hunk 2]
```diff
@@ -58,11 +58,11 @@ class HashedFilesMixin:
                 '//# sourceMappingURL=%(url)s',
             ),
             (
-                r"""(?P<matched>import\s+(?s:(?P<imports>.*?))\s*from\s*["'](?P<url>.*?)["'])""",
+                r"""(?P<matched>import\s+(?s:(?P<imports>[_$a-zA-Z0-9\*\s{},]*?))\s*from\s*["'](?P<url>.*?)["'])""",
                 'import %(imports)s from "%(url)s"',
             ),
             (
-                r"""(?P<matched>export\s+(?s:(?P<exports>.*?))\s*from\s*["'](?P<url>.*?)["'])""",
+                r"""(?P<matched>export\s+(?s:(?P<exports>[_$a-zA-Z0-9\*\s{},]*?))\s*from\s*["'](?P<url>.*?)["'])""",
```
Comment (felixxm) [inline django/contrib/staticfiles/storage.py:65]: As far as I'm concerned special characters lose their special meaning inside sets:
```suggestion
                r"""(?P<matched>export\s+(?s:(?P<exports>[_$a-zA-Z0-9*\s{},]*?))\s*from\s*["'](?P<url>.*?)["'])""",
```

File: django/contrib/staticfiles/storage.py [hunk 3]
```diff
@@ -58,11 +58,11 @@ class HashedFilesMixin:
                 '//# sourceMappingURL=%(url)s',
             ),
             (
-                r"""(?P<matched>import\s+(?s:(?P<imports>.*?))\s*from\s*["'](?P<url>.*?)["'])""",
+                r"""(?P<matched>import\s+(?s:(?P<imports>[_$a-zA-Z0-9\*\s{},]*?))\s*from\s*["'](?P<url>.*?)["'])""",
```
Comment (felixxm) [inline django/contrib/staticfiles/storage.py:61]:
```suggestion
                r"""(?P<matched>import\s+(?s:(?P<imports>[_$a-zA-Z0-9*\s{},]*?))\s*from\s*["'](?P<url>.*?)["'])""",
```

File: tests/staticfiles_tests/project/documents/cached/highlight.pack.js [hunk 4]
```diff
@@ -0,0 +1,2 @@
...eries SeriesCoefficient SeriesData SessionTime Set SetAccuracy SetAlphaChannel SetAttributes Setbacks SetBoxFormNamesPacket SetDelayed SetDirectory SetEnvironment SetEvaluationNotebook SetFileDate SetFileLoadingContext SetNotebookStatusLine SetOptions SetOptionsPacket SetPrecision SetProperty SetSelectedNotebook SetSharedFunction SetSharedVariable SetSpeechParametersPacket SetStreamPosition SetSystemOptions Setter SetterBar SetterBox SetterBoxOptions Setting SetValue Shading Shallow ShannonWavelet ShapiroWilkTest Share Sharpen ShearingMatrix ShearingTransform ShenCastanMatrix Short ShortDownArrow Shortest ShortestMatch ShortestPathFunction ShortLeftArrow ShortRightArrow ShortUpArrow Show ShowAutoStyles ShowCellBracket ShowCellLabel ShowCellTags ShowClosedCellArea ShowContents ShowControls ShowCursorTracker ShowGroupOpenCloseIcon ShowGroupOpener ShowInvisibleCharacters ShowPageBreaks ShowPredictiveInterface ShowSelection ShowShortBoxForm ShowSpecialCharacters ShowStringCharacters ShowSyntaxStyles ShrinkingDelay ShrinkWrapBoundingBox SiegelTheta SiegelTukeyTest Sign Signature SignedRankTest SignificanceLevel SignPadding SignTest SimilarityRules SimpleGraph SimpleGraphQ Simplify Sin S
```
Comment (felixxm) [inline tests/staticfiles_tests/project/documents/cached/highlight.pack.js:1]: We should try to create a minimal :imp: example.

**Review by felixxm — COMMENTED** [review]
@carltongibson Thanks for this patch :+1: I noticed that it doesn't work with type definitions :disappointed: (see my commit) and it crashes with the [highlight.js](https://github.com/highlightjs/highlight.js). We should probably revert the original patch as handling all the different `import`/`export` variants in JavaScript can be too complicated.

File: django/contrib/staticfiles/storage.py (hunk 2 above)
Comment (carltongibson) [reply django/contrib/staticfiles/storage.py:65]: Yes. I went back and forth on that. 👍

File: tests/staticfiles_tests/project/documents/cached/highlight.pack.js (hunk 4 above)
Comment (carltongibson) [reply tests/staticfiles_tests/project/documents/cached/highlight.pack.js:1]: Yes. Just hadn't got there yet.

**Comment by carltongibson** [comment]
Hey @felixxm. Yep. 👍. Maybe we can get this right with a bit more time. 

(Turned out it was "too simple". 😀)

Thanks!

**Comment by felixxm** [comment]
Alternative PR https://github.com/django/django/pull/15064

**Comment by felixxm** [comment]
`export` and `import` directives have many syntax variants and it will be hard to cover them all properly.

**Comment by carltongibson** [comment]
I think we should go with #15064, to at least give time to address this properly (if it's feasible)

**Comment by shaib** [comment]
FWIW @carltongibson @kezabelle https://docs.python.org/3/library/unicodedata.html

**Comment by kezabelle** [comment]
Hey, thanks @shaib :) tbh I only _vaguely_ know of that module, but AFAIK it doesn't offer up the values, à la `string.ascii_letters` or equivalent (which would be I think probably the ideal, if it existed). Somewhat understandably, as I expect just the lookup table to be huge, let alone the contiguous ranges within the various category buckets. Interning the necessary table in Django just to generate a regex is (_one would think_) a big burden even if it were rarely (ever?) modified :/

**Comment by shaib** [comment]
No, I don't think it does, and I also agree that we should not try to duplicate parts of the Javascript parser within Django if they aren't trivial.
We could probably get this right by combining a lax regex with [`isidentifier()`](https://docs.python.org/3/library/stdtypes.html#str.isidentifier), but I think for this issue there are more complications justifying taking the other approach.

**Comment by carltongibson** [comment]
There's a fresh attempt in #16241 — with a slightly more sophisticated regex — if folks here would care to take a look. 

> This works with the higlight.js file but not the type definitions from this review, https://github.com/django/django/pull/15058#pullrequestreview-798503096. Since they are a feature of typescript and not javascript, I think that's a fair trade off.

Whether there's general agreement there would seem to be the question. 

Thanks

**Pull request closed without merging**